### ISSUE TITLE
[rush] rush install should provide an option to skip local linking of dependency projects

### ISSUE THREAD

**Issue by bsiegel** [microsoft/web-build-tools#1385]
Consider the situation where there are two projects in the Rush workspace, and project A depends on project D (for ‘dependency’). Project A and Project D are both at version 1.0.0. Project A depends on `”D”: “^1.0.0”`.

Project D has had some commits made in preparation for an eventual release of version 1.0.1. However it’s not ready to go out the door yet.

Project A has also had some commits, and importantly, these changes do NOT rely on any of the changes in Project D. We are getting ready to publish version 1.0.1 of project A to NPMJS. Thus, we want to build, test, and fully validate a build of Project A version 1.0.1 against the current, published version of D (1.0.0) available on NPMJS.

Today this doesn’t seem possible, because Rush will always link D into Project A’s node_modules. It would be nice to have a flag that can be provided to `rush install` and/or `rush update` to tell it to skip the local linking for a given project (or even for all projects, since that would be simpler).

**Comment by iclanton** [comment]
You can do this by adding a `"cyclicDependencyProjects": [ "D" ]` property to `A`'s configuration entry in `rush.json`.

That will instruct Rush to install `D` from the package feed for `A` instead of locally linking `D`.

**Comment by octogonz** [comment]
We should really implement #547 to avoid this confusion

**Comment by bsiegel** [comment]
@iclanton Yep, I’m aware, but that requires editing the rush.json any time we want to skip linking local projects. Another alternative would be to remove project `D` entirely from rush.json, or to change project `D`’s version temporarily to one that doesn’t match the range that `A` specifies. However all of these would require editing a JSON file. For this issue, the ask is to be able to select at install time whether local linking is done or not.

**Comment by iclanton** [comment]
Running in that mode would require regenerating the shrinkwrap file, which changes the state of the repo as well. It's also not guaranteed to work because there is no way to know if dependencies have been published to a package feed.

There isn't a way to do achieve a non-locally-linked scenario without editing some configuration file - you'll either need to mark dependencies as cyclic for the consuming package or change a hypothetical setting to put the repo in "no-link"-mode, and then regenerate and check-in the shrinkwrap file for that mode.

**Comment by wbern** [comment]
I realize I would want this functionality as well now. I have a case where I want the dev to get up and running without compiling other packages that they depend on in the repo (we have 90 modules).

Now that some PR's are coming in to change common directory, would it be possible to achieve this somehow via a separate common directory for this specific mode?

**Comment by wbern** [comment]
Thinking about it more closely, it really shouldn't be a problem to just download the tar file, and extract it to the package's dist directory..

**Comment by jennysharps** [comment]
With `"useWorkspaces" : "true"`, it looks like when `pnpm` is the package manager, rush sets [`link-workspace-packages=false`](https://pnpm.io/workspaces#link-workspace-packages), which essentially makes the `workspace:` protocol the differentiator between dependencies we want to link in the workspace and dependencies we want to load from the registry (or a "no-link" mode as @iclanton described above).

Using pnpm workspaces, this is what we observed happening in our repo:

When Project A depends on `”D”: “1.0.0”`, and when the last published version of D is `1.0.0`:

1. when D in the current workspace is also at `1.0.0`, A gets `D@1.0.0` from the registry (which is what we expected)

1. when D in the current workspace is out of range, A's install fails with `"A" depends on package "B" (1.0.0) which exists within the workspace but cannot be fulfilled with the specified version range. Either specify a valid version range, or add the package as a cyclic dependency.` (which we did not expect)

If the install hadn't failed with this message, I think pnpm would've installed `B@1.0.0` from the registry just as it had when the current version of B in the workspace _happened to match_ the version we wanted from the registry. If we wanted a workspace package to be installed, we'd explicitly have had to use the `workspace:` protocol in our version specifier, and we hadn't.  I'm thinking this may just be a bug as workspaces are currently experimental?

If there's actually no appetite to provide an option to skip local linking of dependency projects, this seems to conflict with setting `link-workspace-packages` to false when using pnpm workspaces. Since this issue is quite old, can anybody confirm if this thinking has evolved at all in the last 2 years?
